app/coffee/AuthHandler.coffee

```diff
@@ -0,0 +1,85 @@
+_ = require("lodash")
+bcrypt = require("bcrypt")
+crypto = require("crypto")
+logger = require("logger-sharelatex")
+
+{ db } = require("./MongoHandler")
+Errors = require("./Errors")
+
+TOKEN_TTL_MS = 60 * 60 * 1000    # 1 hour
+TOKEN_SEARCH_PREFIX_LENGTH = 16  # 8 bytes
+
+module.exports = AuthHandler =
+	login: (email, password, callback) ->
+		db.users.findOne {email: email.toLowerCase()}, (err, user) ->
+			if err?
+				return callback(err)
+			if !user
+				return callback(new Errors.AuthenticationError("user not found"))
+			if !user.hashedPassword
+				return callback(new Errors.AuthenticationError("user does not have a password"))
+			bcrypt.compare password, user.hashedPassword, (err, match) ->
+				if err?
+					return callback(err)
+				if !match
+					return callback(new Errors.AuthenticationError("incorrect password"))
+				callback(null, user._id.toString())
+
+	oneTimeLogin: (email, token, callback) ->
+		now = new Date()
+		normalizedEmail = email.toLowerCase()
+		tokenPrefix = token.slice(0, TOKEN_SEARCH_PREFIX_LENGTH)
+		tokenPrefixRegexp = new RegExp("^#{_.escapeRegExp(tokenPrefix)}")
+		query = {
+			email: normalizedEmail
+			token: tokenPrefixRegexp
+			expiresAt: { $gt: now }
+			usedAt: { $exists: false }
+		}
+		update = {
+			$set: { usedAt: now }
+		}
+		db.oneTimeLoginTokens.findAndModify { query: query, update: update }, (err, doc) =>
+			if err?
+				return callback(err)
+			if !doc
+				return callback(new Errors.AuthenticationError("invalid token"))
+			if !@_compareTokens(token, doc.token)
+				return callback(new Errors.AuthenticationError("invalid token"))
+			db.users.findOne { email: normalizedEmail }, (err, user) =>
+				if err?
+					return callback(err)
+				if !user
+					return callback(new Errors.AuthenticationError("user not found"))
+				callback(null, user._id.toString())
+
+	generateOneTimeLoginToken: (email, callback) ->
+		db.users.findOne { email: email.toLowerCase() }, (err, user) ->
+			if err?
+				return callback(err)
+			if !user?
+				return callback(new Errors.UserNotFoundError("user not found"))
+
+			logger.info({ email }, "Generating one-time login token")
+			now = new Date()
+			expiresAt = new Date(now.getTime() + TOKEN_TTL_MS)
+			token = crypto.randomBytes(32).toString('hex')
+			doc = {
+				email: email
+				token: token
+				createdAt: now
+				expiresAt: expiresAt
+			}
+			db.oneTimeLoginTokens.insert doc, (err) ->
+				if err?
+					return callback(err)
+				callback(null, token)
+
+	_compareTokens: (token1, token2) ->
+		token1Buffer = Buffer.from(token1)
+		token2Buffer = Buffer.from(token2)
+		if token1Buffer.length != token2Buffer.length
```

Comment:
Maybe it would be simpler to just check the total length of the token as a validation step, before we even query for it?

Reply:
I was thinking that it's easier if the token checking part doesn't assume anything about the length of the token. If, for example, we change the length of the generated token for some reason, the token check works with both versions during the transition.

Reply:
OK, good point. That is a good property to have.